tests/unit/test_charm.py

```diff
@@ -0,0 +1,145 @@
+import unittest
+from glob import glob
+from unittest.mock import MagicMock, Mock, patch
+
+import lightkube
+from lightkube import codecs
+from lightkube.core.exceptions import ApiError
+from ops.model import ActiveStatus, BlockedStatus
+from ops.testing import Harness
+
+from charm import TrainingOperatorCharm
+
+class _FakeResponse:
+    """Used to fake an httpx response during testing only."""
+
+    def __init__(self, code):
+        self.code = code
+
+    def json(self):
+        if self.code == 409:
+            reason = "AlreadyExists"
+        else: reason = ""
+        return {"apiVersion": 1, "code": self.code, "message": "broken", "reason":reason}
+
+class _FakeApiError(ApiError):
+    """Used to simulate an ApiError during testing."""
+
+    def __init__(self, code=400):
+        super().__init__(response=_FakeResponse(code))
+
+@patch("lightkube.core.client.GenericSyncClient", Mock)
+class TestCharm(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(TrainingOperatorCharm)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+
+    # Event handlers
+    @patch("charm.TrainingOperatorCharm._create_crds")
+    @patch("charm.TrainingOperatorCharm._create_resource")
+    def test_install_event(self, create_auth, create_crd):
+        self.harness.charm.on.install.emit()
+
+        # Ensure create_auth and create_crds
+        # are called once
+        create_auth.assert_called_once()
+        create_crd.assert_called_once()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating auth resources
+        create_auth.side_effect = _FakeApiError()
+        self.harness.charm.on.install.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Creating/patching resources failed with code {create_auth.side_effect.response.code}."))
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when creating CRDs
+        create_crd.side_effect = _FakeApiError()
+        self.harness.charm.on.install.emit()
+        self.assertEqual(self.harness.charm.unit.status, BlockedStatus(f"Creating/patching resources failed with code {create_crd.side_effect.response.code}."))
+
+    @patch("charm.TrainingOperatorCharm._update_layer")
+    @patch("charm.TrainingOperatorCharm._patch_resource")
+    def test_config_changed_event(self, patch, update):
+        self.harness.charm.on.config_changed.emit()
+        update.assert_called_once()
+
+        # Run each case of patch here?
+        patch.assert_called()
+
+        # Ensure the unit is in BlockedStatus
+        # on exception when patching auth resources
+        patch.side_effect = _FakeApiError()
```

Comment:
Same comment as above.  This feels like a separate test that can break independently of the first half

Reply:
+1